api: Add export and transcode to Asset and Task

gioelecerati opened the pull request:
<!-------------------------------------------------------------------------
 | Thanks for send a pull request! 🎉
 | First, please make sure you familiar with the contribution guidelines
 | https://github.com/livepeer/livepeer.com/blob/master/CONTRIBUTING.md
 -------------------------------------------------------------------------->

**What does this pull request do? Explain your changes. (required)**

Add export and transcode for the vod flow

**Specific updates (required)**

- Add export and transcode endpoints
- Add TaskResult handler for these events

**How did you test each of these updates (required)**

<!-- A detailed description of how you tested your code changes. Include details of your testing environment, and the tests you ran to see how your change affects other areas of the code, etc. -->

**Does this pull request close any open issues?**

Fixes #894 
Fixes #890 

**Screenshots (optional):**

<!-- Drag some screenshots here, if applicable -->

**Checklist:**

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [X] I have read the **CONTRIBUTING** document.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.

gioelecerati committed:
added export

gioelecerati committed:
added generic handler to handle export

gioelecerati committed:
added scheduleTask - moved pub msgs to scheduleTask

victorges commented on packages/api/src/controllers/asset.ts:
```diff
@@ -169,49 +173,43 @@ app.get("/", authMiddleware({}), async (req, res) => {
   }
-);
+
+  let task = await req.taskScheduler.scheduleTask(asset, "export", {
```
Nice `scheduleTask` helper 😌 

I think the `update` to `phase: "waiting"` should be inside it as well though.

victorges commented on packages/api/src/schema/schema.yaml:
```diff
@@ -1109,6 +1110,24 @@ components:
+                  description: s3 presigned url of the asset to export
+                  example: s3+https://s3.amazonaws.com/livepeer-test/ABC123/filename.mp4
+                ipfs:
+                  type: boolean
```
This **must** be an object, not a boolean, since it will later contain sub-fields. Like for example custom service, ERC1155 metadata, etc.

gioelecerati replied:
Yeah I agree, this was just a placeholder until we know the exact data, but I guess it can be a object placeholder for now

victorges commented on packages/api/src/schema/schema.yaml:
```diff
@@ -1151,6 +1170,22 @@ components:
+                  type: object
+                  $ref: "#/components/schemas/asset"
+                ipfs:
+                  type: boolean
```
I'd go with
```suggestion
                ipfs:
                  type: object
                  properties:
                    videoFileCid:
                      type: string
                      description: IPFS CID of the exported video file
                    erc1155MetadataCid:
                      type: string
                      description: IPFS CID of the default metadata exported for the video
```

custom export has no output I guess. Can you think of any?

gioelecerati replied:
Oh you already defined the IPFS  object for the output! Cool, looks great. Nothing comes to mind right not for the custome export output

victorges commented on packages/api/src/task/task.ts:
```diff
@@ -84,21 +88,65 @@ export default class TaskScheduler {
       }
-      console.log(`task type unknown: ${event.task.type}`);
+
+      if (allowedTasks.includes(event.task.type)) {
```
What is this check for? Shouldn't we update every task with its output even if we seemingly don't support it here? e.g. version migration

If we don't understand what is a given event, we should probably `nack` it instead of processing but ignoring all its information. But I think we should be fine just doing a best effort processing and at least always updating the task object.

gioelecerati replied:
I wanted to add a "completed" status phase in a single place, making sure the task type is allowed, but now that you mention it the task is already created so it's already an allowed type of task. I remove it!

victorges commented on packages/api/src/task/task.ts:
```diff
@@ -84,21 +88,65 @@ export default class TaskScheduler {
-            },
-            output: event.output,
-          });
           return true;
```
Transcode task also generates an output asset. We need a similar logic for it here. (could be related to the confusing `originTaskId` as well)

victorges commented on packages/api/src/task/task.ts:
```diff
@@ -84,21 +88,65 @@ export default class TaskScheduler {
+    });
+    await db.task.update(task.id, {
+      status: { phase: "waiting", updatedAt: Date.now() },
+    });
```
Oh you do already have that here! Please remove the repeated update on every place you call `scheduleTask`.

victorges commented on packages/api/src/task/task.ts:
```diff
@@ -84,21 +88,65 @@ export default class TaskScheduler {
+  ) {
+    let newTask: WithID<Task> = {
+      id: uuid(),
+      name: `asset-upload-${asset.name}-${asset.createdAt}`,
```
Name is inaccurate for the other task types. Either receive it as an argument to the function here or try to use the task type to build the name instead? 

From another perspective, feels like this name is unnecessary, perhaps something we'll just want to compose on the frontend but doesn't have any meaning/necessity in the backend. WDYT of just removing it?

gioelecerati replied:
I agree. Are you thinking about removing the name field or it could be just something like `task-$type-$createdAt`?

victorges commented on packages/www/public/sitemap.xml:
```diff
@@ -1,117 +1,117 @@
-<url><loc>https://livepeer.com/jobs/1412804</loc><changefreq>daily</changefreq><priority>0.7</priority><lastmod>2022-02-08T18:27:53.187Z</lastmod></url>
-<url><loc>https://livepeer.com/jobs/1412803</loc><changefreq>daily</changefreq><priority>0.7</priority><lastmod>2022-02-08T18:27:53.187Z</lastmod></url>
-<url><loc>https://livepeer.com/jobs/1412799</loc><changefreq>daily</changefreq><priority>0.7</priority><lastmod>2022-02-08T18:27:53.187Z</lastmod></url>
+<url><loc>https://livepeer.com</loc><changefreq>daily</changefreq><priority>0.7</priority><lastmod>2022-02-09T14:28:24.457Z</lastmod></url>
```
These changes desn't need to be committed. Just an annoying output from the www build.

gioelecerati committed:
changed ipfs - removed waiting from outside scheduler

gioelecerati committed:
parentAssetId to inputAssetId - outputAssetId

gioelecerati committed:
typo

gioelecerati committed:
fix playbackId length and prefix

gioelecerati committed:
removed task name - remove originTaskId - changed scheduler to support both output and input assets

victorges committed:
schema: Update export task params+output schema

victorges committed:
api: Fixes to the export asset API

E2E tested, it works!!
https://ipfs.io/ipfs/QmRpyvxmxWPUUNSDXodjqDVutsq1ubH9sc85B57Eo7tB75

victorges committed:
api: Add download URL to assets

victorges committed:
api: Add dynamic IPFS URLs to export task outputs

gioelecerati committed:
added ingest to webhook

gioelecerati committed:
Merge branch 'gio/vod/export-transcode' of github.com:livepeer/livepeer-com into gio/vod/export-transcode

gioelecerati committed:
added live-to-vod-asset

gioelecerati committed:
same shardKey as stream - better types on shard schedule

gioelecerati committed:
same shardKey as stream - better types on shard schedule

victorges committed:
api/task: Use our dedicated IPFS gateway

gioelecerati committed:
fixed signed url encoding

gioelecerati committed:
added origin to ingests

gioelecerati committed:
check if asset is ready before export

gioelecerati committed:
check if asset is ready before export

victorges committed:
api/cannon: Remove redundant stream/session fetch

gioelecerati committed:
removed redundant mp4 record url from webhook cannon

gioelecerati committed:
removed baseIngest from cannon

gioelecerati committed:
baseIngest remove

victorges committed:
api/store: Make cleanWriteOnlyResponse recursive

victorges committed:
api: Avoid creating fields that dont exist

victorges committed:
api: Remove transcode task type for now

victorges committed:
api: Fix asset creation on request upload

victorges committed:
api: Clean asset responses as well

victorges committed:
api: Create type for new asset

victorges committed:
api/scheduler: Fix task not getting updated

victorges committed:
api: Do some oneOf magic on export-task-params

victorges committed:
api: Fix asset subfield clean up

victorges committed:
api: Improve task error handling

 - Set asset as failed in case upload fails
 - Do not allow upload if asset has already been uploaded
 - Handle corner cases on task event result procecssing

victorges committed:
api: List assets and tasks by updatedAt desc

victorges committed:
api: Some final nits

victorges committed:
api/task: Set IPFS URLs on regular list as well

victorges committed:
prettier
